### PR TITLE
EM refactoring / EMv2

### PR DESCRIPTION
**Pull request by mkuehbach**
The PR contributing the new em that includes all consolidated readers for em, including EBSD, spectroscopy/EDS/EDXS, nionswift, and plain images

This is a breaking change as em_om, em_spctrscpy, em_nion are about to become obsolete with this PR.

This PR on purpose does not include changes on atom probe tomography as this will be a new separate PR,
appdefs and base classes merged in by the base_class_templates PR #51 from nexus_definitions did not change APM-related classes

### CODE REVIEW

**Comment by mkuehbach** [comment]
Before this PR can be merged, #204 has to be merged, ruff applied, checks done, and respective nexus_definitions PRs merged into fairmat so that the em_refactoring can then use the correct version of the nexus definitions.

**Review by sanbrock — CHANGES_REQUESTED** [review]
- Rebase is needed as nyaml2nxdl has already been taken away.  
- Restructuring is needed to move all debug/development code below the relevant reader subdir.  
- Also note that quite some shared code exists which shall not be duplicated in specific readers, but rather could be collected in a generic util subdir.   
- Note that different reader examples could be grouped and/or harmonised (and then documented) so users can chose which solution they want to follow.  
- Documentation according to the new doc style needs to be added.

**Comment by sherjeelshabih** [comment]
> * Rebase is needed as nyaml2nxdl has already been taken away.
> * Restructuring is needed to move all debug/development code below the relevant reader subdir.
> * Also note that quite some shared code exists which shall not be duplicated in specific readers, but rather could be collected in a generic util subdir.
> * Note that different reader examples could be grouped and/or harmonised (and then documented) so users can chose which solution they want to follow.
> * Documentation according to the new doc style needs to be added.

I would also like to add:
* Test cases that atleast cover the conversion routine.

**Comment by sherjeelshabih** [comment]
@mkuehbach Area D is updating ASE to 3.22.1. Please have a look at this related PR as well: https://github.com/FAIRmat-NFDI/pynxtools/pull/263

File: debug/apex.dev.ipynb
Comment (sherjeelshabih) [inline debug/apex.dev.ipynb:1]: Please remove these Jupyter notebooks. If there is code that can be used for testing, please add it to the test cases.

File: examples/em/image_png_protochips_to_nexus.ods
Comment (sherjeelshabih) [inline examples/em/image_png_protochips_to_nexus.ods:1]: Can these be used for test cases as well?

File: pynxtools/dataconverter/readers/em/concepts/concept_mapper.py [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+"""Utilities for working with NeXus concepts encoded as Python dicts in the concepts dir."""
+
+# pylint: disable=no-member
+
+import pytz
+
+from datetime import datetime
+
+
+def load_from_modifier(terms, fd_dct):
```
Comment (sherjeelshabih) [inline pynxtools/dataconverter/readers/em/concepts/concept_mapper.py:27]: I don't understand what **fd** means. Can you expand the name and drop the _dct part instead?

File: pynxtools/dataconverter/readers/em/concepts/concept_mapper.py [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+            return lst
+    return None
+
+
+def convert_iso8601_modifier(terms, dct: dict):
+    """Implement modifier which transforms nionswift time stamps to proper UTC ISO8601."""
+    if terms is not None:
+        if isinstance(terms, str):
+            if terms in dct.keys():
+                return None
```
Comment (sherjeelshabih) [inline pynxtools/dataconverter/readers/em/concepts/concept_mapper.py:46]: Why is this needed?

File: pynxtools/dataconverter/readers/em/concepts/nxs_object.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+                 value=None,
+                 **kwargs):
+        if (name is not None) and (name == ""):
+            raise ValueError(f"Value for argument name needs to be a non-empty string !")
+        if (unit is not None) and (unit == ""):
+            raise ValueError(f"Value for argument unit needs to be a non-empty string !")
+        if (dtype is not None) and isinstance(dtype, type) is False:
+            raise ValueError(f"Value of argument dtype must not be None "
+                             f" and a valid, ideally a numpy datatype !")
+        # self.doc = None  # docstring
```
Comment (sherjeelshabih) [inline pynxtools/dataconverter/readers/em/concepts/nxs_object.py:43]: Can be removed.

File: pynxtools/dataconverter/readers/em/concepts/nxs_object.py [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+        self.name = name  # name of the field
+        self.unit = unit  # not unit category but actual unit
```
Comment (sherjeelshabih) [inline pynxtools/dataconverter/readers/em/concepts/nxs_object.py:45]: Please use self explanatory variable names wherever possible. It's sloppy to add comments after to explain it when it can be solved much easier with a better name.

File: pynxtools/dataconverter/readers/em/concepts/nxs_object.py [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+        # use special values "unitless" for NX_UNITLESS (e.g. 1) and
+        # "dimensionless" for NX_DIMENSIONLESS (e.g. 1m / 1m)
```
Comment (sherjeelshabih) [inline pynxtools/dataconverter/readers/em/concepts/nxs_object.py:47]: This is not super clear anyways to a new user. You can add these type of docs in the class docstring and list all the arguments. Those docstrings are picked up by doc tools. This won't really surface nicely.

File: pynxtools/dataconverter/readers/em/concepts/nxs_object.py [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+        self.dtype = dtype  # use np.dtype if possible
```
Comment (sherjeelshabih) [inline pynxtools/dataconverter/readers/em/concepts/nxs_object.py:48]: Who is this comment for? Someone using this class? Add such details in class docstrings as well.

File: pynxtools/dataconverter/readers/em/concepts/nxs_object.py [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+        if value is None or isinstance(dtype, str):
+            self.unit = "unitless"
+        self.value = value
+        # value should be a numpy scalar, tensor, or string if possible
```
Comment (sherjeelshabih) [inline pynxtools/dataconverter/readers/em/concepts/nxs_object.py:52]: Same as above.

File: pynxtools/dataconverter/readers/em/concepts/nxs_object.py [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+        self.eqv_hdf = None
```
Comment (sherjeelshabih) [inline pynxtools/dataconverter/readers/em/concepts/nxs_object.py:53]: What is an **eqv**?

File: pynxtools/dataconverter/readers/em/concepts/nxs_spectrum_set.py [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+                                  "spectrum_twod/axis_x@long_name-attribute",
+                                  "spectrum_twod/axis_y-field",
+                                  "spectrum_twod/axis_y@long_name-attribute",
+                                  "spectrum_twod/intensity-field",
+                                  "spectrum_twod/intensity@long_name-attribute"]
+
+
+class NxSpectrumSet():
+    def __init__(self):
+        self.tmp: Dict = {}
```
Comment (sherjeelshabih) [inline pynxtools/dataconverter/readers/em/concepts/nxs_spectrum_set.py:69]: Rename this to something more obvious and explanatory.

File: pynxtools/dataconverter/readers/em/examples/ebsd_database.py [hunk 10]
```diff
@@ -0,0 +1,320 @@
...
+# for the Summer2023 EBSD database example with NOMAD OASIS:
+# we collected several EBSD datasets from colleagues. In these research studies,
+# names of the phases used for indexing (NXem_ebsd_crystal_structure_candidate)
+# instances were very different elements, substances, free-text annotations,
+# mineral names, names of mineral groups, or names of rocks
+# here is the FOR NOW human expert dictionary composed from these names
+# to map used phase names on atom types
+# the idea behind this was to identify what is the minimal information that
+# is recoverable when there is no common agreement about the phases used and their
+# exact atomic configuration
```
Comment (sherjeelshabih) [inline pynxtools/dataconverter/readers/em/examples/ebsd_database.py:29]: Add this to the class docstring if you like. Those can end up in html docs if we need. These won't.

File: pynxtools/dataconverter/readers/em/examples/ebsd_database.py [hunk 11]
```diff
@@ -0,0 +1,320 @@
...
+
+# typical scanning schemes used for EBSD
+# which lattice symmetry
+HEXAGONAL_GRID = "hexagonal_grid"  # typically assuming a tiling with regular hexagons
+SQUARE_GRID = "square_grid"  # a tiling with squares
+REGULAR_TILING = "regular_tiling"
```
Comment (sherjeelshabih) [inline pynxtools/dataconverter/readers/em/examples/ebsd_database.py:35]: These comments can also be avoided with proper naming. For example, **square_grid_tiling** or **scanning_hexagonal**. You will know how best to get this done. But it is always better to have proper var names than #comments explaining them all the time.

File: pynxtools/dataconverter/readers/em/geometry/convention_mapper.py [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+from pynxtools.dataconverter.readers.em.geometry.geometry \
+    import NxEmConventions
+
+# example how to check against different types of Euler angle conventions
+# from pynxtools.dataconverter.readers.em.geometry.euler_angle_convention \
+#    import which_euler_convention
+
+# example how to check if set of conventions matches to some suggestion in the literature
+# from pynxtools.dataconverter.readers.em.geometry.msmse_convention \
+#     import is_consistent_with_msmse_convention
```
Comment (sherjeelshabih) [inline pynxtools/dataconverter/readers/em/geometry/convention_mapper.py:39]: Please place this in a better place and remove it from here.

File: pynxtools/dataconverter/readers/em/geometry/convention_mapper.py [hunk 13]
```diff
@@ -0,0 +1,83 @@
...
+        #     self.entry_id = entry_id
+        #     self.file_name = file_name
+        #     with open(self.file_name, "r", encoding="utf-8") as stream:
+        #         self.yml = fd.FlatDict(yaml.safe_load(stream), delimiter=":")
+        # else:
+        #     self.entry_id = 1
+        #     self.file_name = ""
+        #     self.yml = {}
+        # if "ElectronBackscatterDiffraction" in self.yml:
+        #     self.yml = self.yml["ElectronBackscatterDiffraction"]
```
Comment (sherjeelshabih) [inline pynxtools/dataconverter/readers/em/geometry/convention_mapper.py:66]: remove

File: pynxtools/dataconverter/readers/em/geometry/convention_mapper.py [hunk 14]
```diff
@@ -0,0 +1,83 @@
...
+                res = value
+                if res is not None:
+                    template[trg] = res
+        return template
+        # self.parse_rotation_convention_section(template)
+        # self.parse_processing_frame_section(template)
+        # self.parse_sample_frame_section(template)
+        # self.parse_detector_frame_section(template)
+        # self.parse_gnomonic_projection_section(template)
+        # return template
```
Comment (sherjeelshabih) [inline pynxtools/dataconverter/readers/em/geometry/convention_mapper.py:83]: Remove as well.

File: pynxtools/dataconverter/readers/em/geometry/euler_angle_convention.py [hunk 15]
```diff
@@ -0,0 +1,63 @@
...
+# described by an attached Cartesian CS named B
+# the sequence about which specific axis and new/intermediate axes
+# one rotates enables to distinguish different types of Euler angle
+# so-called Euler-angle conventions, most commonly used in materials
+# science is the convention of H.-J. Bunge aka zxz convention
+
+# "Bunge" https://doi.org/10.1016/C2013-0-11769-2
+# "Rowenhorst" https://doi.org/10.1088/0965-0393/23/8/083501
+# "Morawiec" https://doi.org/10.1007/978-3-662-09156-2
+# "Britton" https://doi.org/10.1016/j.matchar.2016.04.008
```
Comment (sherjeelshabih) [inline pynxtools/dataconverter/readers/em/geometry/euler_angle_convention.py:34]: Move to the class docstring for better exposure.

File: pynxtools/dataconverter/readers/em/geometry/euler_angle_convention.py [hunk 16]
```diff
@@ -0,0 +1,63 @@
...
+
+
+which_euler_convention = {"xxx": (False, "", ""),
```
Comment (sherjeelshabih) [inline pynxtools/dataconverter/readers/em/geometry/euler_angle_convention.py:37]: You could drop the **which_** prefix. Or add a verb to it. This reads a bit off.

File: pynxtools/dataconverter/readers/em/geometry/handed_cartesian.py [hunk 17]
```diff
@@ -0,0 +1,749 @@
...
+# and thus unique definition. All other choices are reported as a not fully constraint
+# i.e. eventually problematic coordinate system
+
+REFERENCE_FRAMES = ["undefined", "right_handed_cartesian", "left_handed_cartesian"]
+
+AXIS_DIRECTIONS = ["undefined", "north", "east", "south", "west", "in", "out"]
+
+# is a right-handed (rh) Cartesian coordinate system (cs) sufficiently constrained, when
+# at least two base vector directions are chosen, the order is x, y, z
+is_cs_rh_unambiguous = {"undefined_undefined_undefined": False,
```
Comment (sherjeelshabih) [inline pynxtools/dataconverter/readers/em/geometry/handed_cartesian.py:48]: You're using dicts almost as functions. This is quite an irregular syntax. Consider using lamba functions or the sorts for this if you wanna save space.

File: pynxtools/dataconverter/readers/em/geometry/msmse_convention.py [hunk 18]
```diff
@@ -0,0 +1,51 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+"""Conventions used in the EBSD communnity
+
+suggested by D. Rowenhorst et al. in DOI: 10.1088/0965-0393/23/8/083501.
+"""
```
Comment (sherjeelshabih) [inline pynxtools/dataconverter/readers/em/geometry/msmse_convention.py:21]: This looks like a mistaken newline character. You can remove this so it reads pretty for HTML as well.

File: pynxtools/dataconverter/readers/em/geometry/msmse_convention.py [hunk 19]
```diff
@@ -0,0 +1,51 @@
...
+        "axis_angle_convention"]
+    for field_name in req_fields:
+        if (field_name not in dct) or (field_name not in msmse_convention):
+            return "unclear"
+    # okay, so either matching or not
+    for field_name in req_fields:
+        if dct[field_name] == msmse_convention[field_name]:
+            continue
+        return "inconsistent"
+    return "consistent"
```
Comment (sherjeelshabih) [inline pynxtools/dataconverter/readers/em/geometry/msmse_convention.py:51]: I don't fully understand this. Could it be done in one loop?

File: pynxtools/dataconverter/readers/em/reader.py [hunk 20]
```diff
@@ -0,0 +1,185 @@
...
+# from pynxtools.dataconverter.readers.em_om.utils.orix_ebsd_parser \
+#     import NxEmOmOrixEbsdParser
+# from pynxtools.dataconverter.readers.em_om.utils.mtex_ebsd_parser \
+#     import NxEmOmMtexEbsdParser
+# from pynxtools.dataconverter.readers.em_om.utils.zip_ebsd_parser \
+#     import NxEmOmZipEbsdParser
+# from pynxtools.dataconverter.readers.em_om.utils.dream3d_ebsd_parser \
+#     import NxEmOmDreamThreedEbsdParser
+# from pynxtools.dataconverter.readers.em_om.utils.em_nexus_plots \
+#     import em_om_default_plot_generator
```
Comment (sherjeelshabih) [inline pynxtools/dataconverter/readers/em/reader.py:46]: Please remove unused code.

File: pynxtools/dataconverter/readers/em/reader.py [hunk 21]
```diff
@@ -0,0 +1,185 @@
...
+        debugging = False
+        if debugging is True:
+            print("Reporting state of template before passing to HDF5 writing...")
+            for keyword in template.keys():
+                print(keyword)
+                # print(type(template[keyword]))
+                print(template[keyword])
+
+        print("Forward instantiated template to the NXS writer...")
+        return template
```
Comment (sherjeelshabih) [inline pynxtools/dataconverter/readers/em/reader.py:181]: Unused code.

**Review by sherjeelshabih — CHANGES_REQUESTED** [review]
I've done a partial review till now. There are many repeating things. Can you please clear the pull request and keep only what's important? Maybe also apply the suggestions up till now to the rest of the parts and then I can have another look.

File: debug/apex.dev.ipynb
Comment (mkuehbach) [reply debug/apex.dev.ipynb:1]: no longer present in pynxtools-em

File: pynxtools/dataconverter/readers/em/reader.py (hunk 21 above)
Comment (mkuehbach) [reply pynxtools/dataconverter/readers/em/reader.py:181]: unused because was deactivated intentionally for debugging purposes, cleaned in pynxtools-em

File: pynxtools/dataconverter/readers/em/geometry/msmse_convention.py (hunk 19 above)
Comment (mkuehbach) [reply pynxtools/dataconverter/readers/em/geometry/msmse_convention.py:51]: Yes, was refactored in pynxtools-em

File: pynxtools/dataconverter/readers/em/reader.py (hunk 20 above)
Comment (mkuehbach) [reply pynxtools/dataconverter/readers/em/reader.py:46]: intentionally deactivated for debugging but indeed removed or reactivated in pynxtools-em

File: pynxtools/dataconverter/readers/em/geometry/msmse_convention.py (hunk 18 above)
Comment (mkuehbach) [reply pynxtools/dataconverter/readers/em/geometry/msmse_convention.py:21]: fixed in pynxtools-em

File: pynxtools/dataconverter/readers/em/geometry/convention_mapper.py (hunk 13 above)
Comment (mkuehbach) [reply pynxtools/dataconverter/readers/em/geometry/convention_mapper.py:66]: intentionally deactivated for debugging but indeed removed or reactivated in pynxtools-em

File: pynxtools/dataconverter/readers/em/geometry/convention_mapper.py (hunk 14 above)
Comment (mkuehbach) [reply pynxtools/dataconverter/readers/em/geometry/convention_mapper.py:83]: intentionally deactivated for debugging but indeed removed or reactivated in pynxtools-em

File: pynxtools/dataconverter/readers/em/geometry/euler_angle_convention.py (hunk 16 above)
Comment (mkuehbach) [reply pynxtools/dataconverter/readers/em/geometry/euler_angle_convention.py:37]: dropped which

File: pynxtools/dataconverter/readers/em/geometry/convention_mapper.py (hunk 12 above)
Comment (mkuehbach) [reply pynxtools/dataconverter/readers/em/geometry/convention_mapper.py:39]: removed in pynxtools-em

File: pynxtools/dataconverter/readers/em/examples/ebsd_database.py (hunk 11 above)
Comment (mkuehbach) [reply pynxtools/dataconverter/readers/em/examples/ebsd_database.py:35]: agreed, fixed in pynxtools-em

File: pynxtools/dataconverter/readers/em/examples/ebsd_database.py (hunk 10 above)
Comment (mkuehbach) [reply pynxtools/dataconverter/readers/em/examples/ebsd_database.py:29]: did so in pynxtools-em, also in those places commented here

File: pynxtools/dataconverter/readers/em/concepts/nxs_object.py [hunk 22]
```diff
@@ -0,0 +1,62 @@
...
+
+
+class NxObject:
+    """An object in a graph e.g. an attribute, dataset, or group in NeXus."""
+
+    def __init__(self,
+                 name: str = None,
+                 unit: str = None,
+                 dtype=None,
+                 value=None,
```
[the remark this answers is not in the document]
Comment (mkuehbach) [reply pynxtools/dataconverter/readers/em/concepts/nxs_object.py:34]: removed in pynxtools-em

File: pynxtools/dataconverter/readers/em/concepts/nxs_object.py (hunk 4 above)
Comment (mkuehbach) [reply pynxtools/dataconverter/readers/em/concepts/nxs_object.py:45]: fixed in pynxtools-em

File: pynxtools/dataconverter/readers/em/concepts/nxs_object.py (hunk 5 above)
Comment (mkuehbach) [reply pynxtools/dataconverter/readers/em/concepts/nxs_object.py:47]: fixed in pynxtools-em

File: pynxtools/dataconverter/readers/em/concepts/nxs_object.py (hunk 6 above)
Comment (mkuehbach) [reply pynxtools/dataconverter/readers/em/concepts/nxs_object.py:48]: moved to docstring as suggested in pynxtools-em

File: pynxtools/dataconverter/readers/em/concepts/nxs_object.py (hunk 7 above)
Comment (mkuehbach) [reply pynxtools/dataconverter/readers/em/concepts/nxs_object.py:52]: fixed as above

File: pynxtools/dataconverter/readers/em/concepts/nxs_object.py (hunk 8 above)
Comment (mkuehbach) [reply pynxtools/dataconverter/readers/em/concepts/nxs_object.py:53]: explained in docstring

File: pynxtools/dataconverter/readers/em/concepts/concept_mapper.py [hunk 23]
```diff
@@ -0,0 +1,116 @@
...
+# print(apply_modifier(modd, yml))
+
+def variadic_path_to_specific_path(path: str, instance_identifier: list):
+    """Transforms a variadic path to an actual path with instances."""
+    if (path is not None) and (path != ""):
+        narguments = path.count("*")
+        if narguments == 0:  # path is not variadic
+            return path
+        if len(instance_identifier) >= narguments:
+            tmp = path.split("*")
```
[the remark this answers is not in the document]
Comment (mkuehbach) [reply pynxtools/dataconverter/readers/em/concepts/concept_mapper.py:108]: implemented as suggested in pynxtools-em

File: pynxtools/dataconverter/readers/em/concepts/concept_mapper.py [hunk 24]
```diff
@@ -0,0 +1,116 @@
...
+            if len(tmp) == narguments + 1:
+                nx_specific_path = ""
+                for idx in range(0, narguments):
+                    nx_specific_path += f"{tmp[idx]}{instance_identifier[idx]}"
+                    idx += 1
+                nx_specific_path += f"{tmp[-1]}"
+                return nx_specific_path
+    return None
```
[the remark this answers is not in the document]
Comment (mkuehbach) [reply pynxtools/dataconverter/readers/em/concepts/concept_mapper.py:116]: fixed in pynxtools-em

File: pynxtools/dataconverter/readers/em/concepts/concept_mapper.py [hunk 25]
```diff
@@ -0,0 +1,116 @@
...
+# modd = {"fun": "load_from", "terms": ["metadata/scan/scan_device_properties/mag_boards/MagBoard 1 DAC 11",
+#     "metadata/scan/scan_device_properties/mag_boards/MagBoard 1 Relay"]}
+# modd = {"fun": "convert_iso8601", "terms": ["data_modified", "timezone"]}
+# print(apply_modifier(modd, yml))
+
+def variadic_path_to_specific_path(path: str, instance_identifier: list):
+    """Transforms a variadic path to an actual path with instances."""
+    if (path is not None) and (path != ""):
+        narguments = path.count("*")
+        if narguments == 0:  # path is not variadic
```
[the remark this answers is not in the document]
Comment (mkuehbach) [reply pynxtools/dataconverter/readers/em/concepts/concept_mapper.py:105]: implemented as suggested

File: pynxtools/dataconverter/readers/em/concepts/concept_mapper.py [hunk 26]
```diff
@@ -0,0 +1,116 @@
...
+                        = datetime.strptime(date_time_str, '%Y-%m-%d %H:%M:%S.%f')
+                    utc_time_zone_aware \
+                        = pytz.timezone(time_zone_str).localize(date_time_obj)
+                    return utc_time_zone_aware
+                else:
+                    raise ValueError('Invalid timezone string!')
+                return None
+        else:
+            return None
+    return None
```
[the remark this answers is not in the document]
Comment (mkuehbach) [reply pynxtools/dataconverter/readers/em/concepts/concept_mapper.py:69]: changed as suggested in pynxtools-em

File: pynxtools/dataconverter/readers/em/concepts/concept_mapper.py [hunk 27]
```diff
@@ -0,0 +1,116 @@
...
+
+
+def apply_modifier(modifier, dct: dict):
+    """Interpret a functional mapping using data from dct via calling modifiers."""
+    if isinstance(modifier, dict):
+        # different commands are available
```
[the remark this answers is not in the document]
Comment (mkuehbach) [reply pynxtools/dataconverter/readers/em/concepts/concept_mapper.py:75]: removed in pynxtools-em

File: pynxtools/dataconverter/readers/em/concepts/concept_mapper.py [hunk 28]
```diff
@@ -0,0 +1,116 @@
...
+                return convert_iso8601_modifier(modifier["terms"], dct)
+        elif set(["link"]) == set(modifier.keys()):
+            # CURRENTLY NOT IMPLEMENTED
+            # with the jsonmap reader Sherjeel conceptualized "link"
+            return None
+        else:
+            return None
+    if isinstance(modifier, str):
+        return modifier
+    return None
```
[the remark this answers is not in the document]
Comment (mkuehbach) [reply pynxtools/dataconverter/readers/em/concepts/concept_mapper.py:89]: implemented as suggested in pynxtools-em

File: pynxtools/dataconverter/readers/em/concepts/concept_mapper.py (hunk 2 above)
Comment (mkuehbach) [reply pynxtools/dataconverter/readers/em/concepts/concept_mapper.py:46]: need to check, have noted this down but resolvable

File: pynxtools/dataconverter/readers/em/concepts/concept_mapper.py [hunk 29]
```diff
@@ -0,0 +1,116 @@
...
+def convert_iso8601_modifier(terms, dct: dict):
+    """Implement modifier which transforms nionswift time stamps to proper UTC ISO8601."""
+    if terms is not None:
+        if isinstance(terms, str):
+            if terms in dct.keys():
+                return None
+        elif (isinstance(terms, list)) and (len(terms) == 2) \
+                and (all(isinstance(entry, str) for entry in terms) is True):
+            # assume the first argument is a local time
+            # assume the second argument is a timezone string
```
[the remark this answers is not in the document]
Comment (mkuehbach) [reply pynxtools/dataconverter/readers/em/concepts/concept_mapper.py:50]: fixed in pynxtools-em

File: pynxtools/dataconverter/readers/em/concepts/concept_mapper.py [hunk 30]
```diff
@@ -0,0 +1,116 @@
...
+
+
+# examples/tests how to use modifiers
+# modd = "µs"
+# modd = {"link": "some_link_to_somewhere"}
+# modd = {"fun": "load_from", "terms": "metadata/scan/scan_device_properties/mag_boards/MagBoard 1 DAC 11"}
+# modd = {"fun": "load_from", "terms": ["metadata/scan/scan_device_properties/mag_boards/MagBoard 1 DAC 11",
+#     "metadata/scan/scan_device_properties/mag_boards/MagBoard 1 Relay"]}
+# modd = {"fun": "convert_iso8601", "terms": ["data_modified", "timezone"]}
+# print(apply_modifier(modd, yml))
```
[the remark this answers is not in the document]
Comment (mkuehbach) [reply pynxtools/dataconverter/readers/em/concepts/concept_mapper.py:99]: it is, has been removed in pynxtools-em

File: pynxtools/dataconverter/readers/em/concepts/concept_mapper.py (hunk 1 above)
Comment (mkuehbach) [reply pynxtools/dataconverter/readers/em/concepts/concept_mapper.py:27]: changed based on suggestion in pynxtools-em

File: examples/em/image_png_protochips_to_nexus.ods
Comment (mkuehbach) [reply examples/em/image_png_protochips_to_nexus.ods:1]: no, was for dev, already removed in pynxtools-em

File: pynxtools/dataconverter/readers/em/geometry/handed_cartesian.py (hunk 17 above)
Comment (mkuehbach) [reply pynxtools/dataconverter/readers/em/geometry/handed_cartesian.py:48]: This is a look-up table

**Comment by mkuehbach** [comment]
Closed as all development has been migrated to [pynxtools-em, ](https://github.com/FAIRmat-NFDI/pynxtools-em) remaining suggestions from @sherjeelshabih have been resolved in that repo
